File: PCM.py
import numpy as np
import matplotlib.pyplot as plt

# Parameters
fm = 3
fs = 30
t = np.arange(0, 1, 0.001/fm)
Am = 2
n = 3
L = 2 ** n
ts = np.arange(0, 1, 1/fs)

# Message signal
x_t = Am * np.sin(2 * np.pi * fm * t)

# Sampled signal
x_s = Am * np.sin(2 * np.pi * fm * ts)

# Quantization parameters
Vmin = -Am
Vmax = Am
delta = (Vmax - Vmin) / (L - 1)
delta_steps = np.arange(Vmin, Vmax + delta, delta)
quant = np.arange(Vmin - delta/2, Vmax + delta/2, delta)

# Quantization
ind = np.digitize(x_s, delta_steps) - 1
ind[ind < 0] = 0
ind[ind >= L] = L - 1  # Ensure indices do not exceed range
x_q = quant[ind]

# Encoding signal
encoded_signal = np.array([np.binary_repr(i, width=n) for i in ind])

# Flatten encoded bits for plotting
encoded_bits = np.array([int(bit) for bits in encoded_signal for bit in bits])

# Plotting
plt.figure(figsize=(10, 10))

plt.subplot(4, 1, 1)
plt.plot(t, x_t, 'k-', linewidth=1.5, label='Original Signal')
plt.stem(ts, x_s, 'g-', label='Sampled Signal', use_line_collection=True)
plt.title('Sampling')
plt.xlabel('Time t (in sec)')
plt.ylabel('Amplitude')
plt.legend()
plt.grid()

plt.subplot(4, 1, 2)
plt.stem(ts, x_s, 'go', use_line_collection=True, label='Sampled values')
plt.stem(ts, x_q, 'r.', use_line_collection=True, label='Quantized values')
plt.title('Sampled vs Quantized')
plt.xlabel('Time t (in sec)')
plt.ylabel('Amplitude')
plt.legend()
plt.grid()

plt.subplot(4, 1, 3)
plt.step(ts, x_q, 'r--', linewidth=1.5, label='Quantized Signal')
plt.stem(ts, x_s, 'go', use_line_collection=True, label='Sampled Signal')
plt.title('Quantization')
plt.xlabel('Time t (in sec)')
plt.ylabel('Amplitude')
plt.legend()
plt.grid()

plt.subplot(4, 1, 4)
plt.step(np.arange(len(encoded_bits)), encoded_bits, 'b-', linewidth=1.5, label='Encoded digital signal')
plt.title('Encoding')
plt.xlabel('Encoded bits')
plt.ylabel('Binary Signal')
plt.grid()

plt.tight_layout()
plt.show()

# Calculating parameters
Rb = fs * n
Tb = 1 / Rb
Qe_max = delta / 2
P_signal = np.mean(x_t ** 2)
P_signal_dB = 10 * np.log10(P_signal)
P_noise = delta ** 2 / 12
P_noise_dB = 10 * np.log10(P_noise)
SQNR_dB = P_signal_dB - P_noise_dB
SQNR_formula_dB = 6.02 * n + 1.76

# Display results
print(f'Bit rate (Rb): {Rb:.2f} bits/sec')
print(f'Bit duration (Tb): {Tb:.2e} sec')
print(f'Step size (δ): {delta:.4f} volts')
print(f'Maximum quantization error (Qe_max): {Qe_max:.4f} volts')
print(f'Signal power: {P_signal_dB:.2f} dB')
print(f'Quantization noise power: {P_noise_dB:.2f} dB')
print(f'SQNR (calculated): {SQNR_dB:.2f} dB')
print(f'SQNR (formula): {SQNR_formula_dB:.2f} dB')

File: PCM.m
fm = 3;
fs = 30;
t = 0:0.001/fm:1;
Am = 2;
n = 3;
L = 2^n;
ts = 0:1/fs:1;

% Message signal
x_t = Am * sin(2 * pi * fm * t);

% Sampled signal
x_s = Am * sin(2 * pi * fm * ts);

% Quantization parameters
Vmin = -Am;
Vmax = Am;
delta = (Vmax - Vmin) / (L - 1);
delta_steps = Vmin:delta:Vmax;
quant = (Vmin - delta/2):delta:(Vmax + delta/2);

% Quantization
[ind, x_q] = quantiz(x_s, delta_steps, quant);
NonZeroInd = find(ind ~= 0);
ind(NonZeroInd) = ind(NonZeroInd) - 1;
encoded_signal = de2bi(ind,n, 'left-msb');

figure;
subplot(4, 1, 1);
plot(t, x_t, 'k-', 'LineWidth', 1.5); % Original signal in black
hold on;
stem(ts, x_s, 'g', 'LineWidth', 1.5); % Sampled signal in green
title('Sampling');
xlabel('Time t (in sec)');
ylabel('Amplitude');
legend('Original Signal', 'Sampled Signal');
grid on;

subplot(4, 1, 2);
stem(ts, x_s, 'go', 'LineWidth', 1.5); % Sampled values in green
hold on;
stem(ts, x_q, 'r.', 'LineWidth', 1.5); % Quantized values in red
title('Sampled vs Quantized');
xlabel('Time t (in sec)');
ylabel('Amplitude');
legend('Sampled values', 'Quantized values');
grid on;

subplot(4, 1, 3);
stairs(ts, x_q, 'r--', 'LineWidth', 1.5); % Quantized signal in red dashed line
hold on;
stem(ts, x_s, 'go', 'LineWidth', 1.5); % Sampled signal in green
title('Quantization');
xlabel('Time t (in sec)');
ylabel('Amplitude');
legend('Quantized Signal', 'Sampled Signal');
grid on;

subplot(4, 1, 4);
encoded_bits = reshape(encoded_signal', 1, []);
stairs(1:length(encoded_bits), encoded_bits, 'b-', 'LineWidth', 1.5); % Encoded digital signal in blue
title('Encoding');
xlabel('Encoded bits');
ylabel('Binary Signal');
grid on;

Rb = fs * n;
Tb = 1 / Rb;
delta = (Vmax - Vmin) / (L - 1);
Qe_max = delta / 2;
P_signal = mean(x_t.^2);
P_signal_dB = 10 * log10(P_signal);
P_noise = delta^2 / 12;
P_noise_dB = 10 * log10(P_noise);
SQNR_dB = P_signal_dB - P_noise_dB;
SQNR_formula_dB = 6.02 * n + 1.76;

% Display results
fprintf('Bit rate (Rb): %.2f bits/sec\n', Rb);
fprintf('Bit duration (Tb): %.2e sec\n', Tb);
fprintf('Step size (δ): %.4f volts\n', delta);
fprintf('Maximum quantization error (Qe_max): %.4f volts\n', Qe_max);
fprintf('Signal power: %.2f dB\n', P_signal_dB);
fprintf('Quantization noise power: %.2f dB\n', P_noise_dB);
fprintf('SQNR (calculated): %.2f dB\n', SQNR_dB);
fprintf('SQNR (formula): %.2f dB\n', SQNR_formula_dB);
